File: subway.js
var express = require('express');
var app = express();
var bodyParser = require('body-parser')
var changeCase = require("change-case");

fs = require('fs')

app.use(bodyParser.json()); // for parsing application/json
app.use(bodyParser.urlencoded({ extended: true })); // for parsing application/x-www-form-urlencoded

app.post('/subway', function (req, res) {

  fs.readFile('serviceStatus.txt', 'utf8', function (err,data) {
    var parseString = require('xml2js').parseString;

    parseString(data, function (err, result) {
      //console.log(result);
      if (err) {
        return console.log(err);
      }

      if (req.body.request.intent != undefined) {
        console.log("query: ",req.body.request.intent.name);
      }
  
      console.log("Got request: %o", req.body);

      say_this = "";

      if (req.body.request.intent == undefined || req.body.request.intent.name == "GetAllStatus") {
        for (var i=0; i<result["service"]["subway"][0]["line"].length;i++) {
          line = result["service"]["subway"][0]["line"][i];
          say_this = say_this.concat(line["name"][0].replace(/(\S{1})/g,"$1 "))
          say_this = say_this.concat("line has ")
          say_this = say_this.concat(line["status"][0])
          say_this = say_this.concat(", ")
        }

      } else if (req.body.request.intent.name == "GetStatus") {
	line_suspected = changeCase.upperCase(req.body.request.intent.slots.Line.value).replace(/ /g,'') //uppercase, remove whitespace
        
        console.log("BINGO:", req.body.request.intent.slots)
        for (var i=0; i<result["service"]["subway"][0]["line"].length;i++) {
          line = result["service"]["subway"][0]["line"][i]
	  line_name = line["name"][0]
          if (line_suspected == "ONE") line_suspected = 1
          if (line_suspected == "TWO") line_suspected = 2
          if (line_suspected == "THREE") line_suspected = 3
          if (line_suspected == "FOUR" || line_suspected == "FOR") line_suspected = 4
          if (line_suspected == "FIVE") line_suspected = 5
          if (line_suspected == "SIX") line_suspected = 6
          if (line_suspected == "SEVEN") line_suspected = 7
          if (line_suspected == "EIGHT") line_suspected = 8
          if (line_suspected == "NINE") line_suspected = 9 
            console.log("comparing", line_suspected, "with", line_name)
          if (line_name.indexOf(line_suspected) > -1) {
            say_this = say_this.concat("The ")
            say_this = say_this.concat(line_name.replace(/(\S{1})/g,"$1 "))
            say_this = say_this.concat("line has ")
            say_this = say_this.concat(line["status"][0])
            say_this = say_this.concat(", ")
          } 
        }

	if (say_this == "")
          say_this = "I'm not sure what subway line you're asking about. At the moment, I can not understand letters. I can only understand numbers. Please put in a feature request for Alexa to understand letters."
      } 

      console.log("say this is:", say_this)

    text = {"outputSpeech": {"type": "PlainText", "text": say_this}}
    response = { 
      "version": "0.1", 
//    "sessionAttributes": { "someInfoOnPreviousRequest": { "last_line": "something"   } },
      "response": text,
      "shouldEndSession": true  
    } 

    console.log("Sending response: ", JSON.stringify(response))
    res.send(JSON.stringify(response));
 
    });
  });
});

var server = app.listen(4445, function () {
  var host = server.address().address;
  var port = server.address().port;

  console.log('alexa-nyc-mta skill listening at http://%s:%s', host, port);
});

File: subway.rb
# subway.rb - alexa nyc mta subway 
#

require 'sinatra'
require 'nokogiri'
require 'json'

require_relative 'lib/alexa_nyc_mta'

def lines_from_state(lines) 
   line_text = String.new
   # for line in delayed_lines do
   lines.each_with_index do |line, idx|
     if (idx != (lines.count-1))
        line_text << line.gsub(/(.{1})/, '\1 ')
        if (idx != lines.count-2) then line_text << ", " end
        puts "idx is #{idx} lines count is #{lines.count}"
        puts line
     else
        line_text << " and the " + line.gsub(/(.{1})/, '\1 ') 
        puts line
      end
   end
   line_text
end


post '/subway/v0.1' do 
   begin
       subway = AlexaNYCMTA.new 
       status = subway.subway_status
       
       request_body = JSON.parse request.body.read
       puts "request body is #{request_body}"
#       puts "subway #{status}"
       intent = request_body["request"]["intent"]["name"] if request_body["request"]["intent"]
       puts "intent #{intent}"
       say_this = ""
       if (intent == "GetAllStatus" || intent == nil)
          good_lines = status.select { |key, value| value.match("GOOD SERVICE")}.transpose[0]
          delayed_lines = status.select { |key, value| value.match("DELAYS")}.transpose[0]
          planned_work_lines = status.select { |key, value| value.match("PLANNED WORK")}.transpose[0]
          service_change_lines = status.select { |key, value| value.match("SERVICE CHANGE")}.transpose[0]
          puts "good lines are #{good_lines}"
          if (delayed_lines || planned_work_lines || service_change_lines)
             say_this = "ok. here's the situation... "
             if (delayed_lines && delayed_lines.count > 0)
                if (delayed_lines.count == 1)
                   delayed_line = delayed_lines[0].gsub(/(.{1})/, '\1 ')
                   say_this << "The #{delayed_line} is currently experiencing delays. " 
                else
                   say_this << "The #{lines_from_state(delayed_lines)} are currently experiencing delays.  " 
                end
             end
             if (planned_work_lines && planned_work_lines.count > 0)
                puts "planned_work_lines are #{planned_work_lines}"
                if (planned_work_lines.count == 1)
                   planned_work_line = planned_work_lines[0].gsub(/(.{1})/, '\1 ')
                   say_this << "The #{planned_work_line} is currently undergoing planned work.  " 
                else
                   say_this << "The #{lines_from_state(planned_work_lines)} are currently undergoing planned work.  " 
                end
             end
             if (service_change_lines && service_change_lines.count > 0)
                if (service_change_lines.count == 1)
                   service_change_line = service_change_lines[0].gsub(/(.{1})/, '\1 ')
                   say_this << "The #{service_change_line} has a major service change.  " 
                else
                   say_this << "The #{lines_from_state(service_change_lines)} have major service changes.  " 
                end
             end

             say_this << "All other lines are currently running smoothly."
          else
             say_this = "ok. Great news. "
             say_this << "All lines are currently running smoothly."
          end
          
          #say_this = "The #{delayed_lines}" 
       else
     
         say_this = "I can't do status of individual lines yet. Please check back in a few days."
         
       end

#   rescue
#      say_this = "rut roh. something failed"

   end 
   text = {"outputSpeech" => {"type" => "PlainText", "text" => say_this}}
   response = {
       "version" => "0.2",
       "sessionAttributes" => { "someInfoOnPreviousRequest" => { "last_line" => "something" } },
        "response" => text,
        "shouldEndSession" => true
   } 
 
   puts "response: #{response.to_json}"
 
   response.to_json
end


